frontend/src/lib/Arrow.ts

```diff
@@ -0,0 +1,288 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Column, Table } from "apache-arrow"
+import { range, unzip } from "lodash"
+
+import { IStyler } from "autogen/proto"
+
+interface Schema {
+  index_columns: (string | RangeIndex)[]
+  columns: SchemaColumn[]
+  column_indexes: any[]
+}
+
+interface RangeIndex {
+  kind: "range"
+  name: string | null
+  start: number
+  step: number
+  stop: number
+}
+
+interface SchemaColumn {
+  field_name: string
+  metadata: Record<string, any> | null
+  name: string | null
+  numpy_type: string
+  pandas_type: string
+}
+
+interface TableDimensions {
+  headerRows: number
+  headerColumns: number
+  dataRows: number
+  dataColumns: number
+  rows: number
+  columns: number
+}
+
+type TableCellType = "blank" | "index" | "columns" | "data"
+
+interface TableCell {
+  type: TableCellType
+  id?: string
+  classNames: string
+  content: string
+}
+
+export interface Styler {
+  uuid?: string
+  caption?: string
+  styles?: string
+  displayValues?: Arrow
+}
+
+export class Arrow {
+  private table: Table
+
+  private schema: Schema
+
+  private index: string[][]
+
+  private columns: string[][]
+
+  private data: any[][]
+
+  private styler?: Styler
+
+  constructor(dataBuffer: Uint8Array, styler?: IStyler) {
+    this.table = Table.from(dataBuffer)
+    this.schema = this.getSchema()
+    this.index = this.getIndex()
+    this.columns = this.getColumns()
+    this.data = this.getData()
+    this.styler = {
+      uuid: styler?.uuid || undefined,
+      caption: styler?.caption || undefined,
+      styles: styler?.styles || undefined,
+      displayValues: styler?.displayValues
+        ? new Arrow(styler.displayValues)
+        : undefined,
+    }
+  }
+
+  private getSchema(): Schema {
+    const schema = this.table.schema.metadata.get("pandas")
+    if (schema == null) {
+      throw new Error("Table schema is missing.")
+    }
+    return JSON.parse(schema)
+  }
+
+  private static getColumnData(column: Column): any[] {
+    return range(0, column.length).map(rowIndex => column.get(rowIndex))
+  }
+
+  private getIndex(): string[][] {
+    return this.schema.index_columns.map(field => {
+      const isRangeIndex = typeof field === "object" && field.kind === "range"
+      if (isRangeIndex) {
+        const { start, stop, step } = field as RangeIndex
+        return range(start, stop, step).map(String)
+      }
+      const column = this.table.getColumn(field as string)
+      return Arrow.getColumnData(column).map(String)
+    })
+  }
+
+  private getColumns(): string[][] {
+    const isMultiIndex = this.schema.column_indexes.length > 1
+    return unzip(
+      this.schema.columns
+        .map(column => column.field_name)
```

Comment:
You don't really need this map. I assume columns should not get expensive so it's probably fine